mypy/checker.py

```diff
@@ -488,16 +235,23 @@ def accept(self, node: Node, type_context: Type = None) -> Type:
         else:
             return typ
 
-    def accept_in_frame(self, node: Node, type_context: Type = None,
-                        repeat_till_fixed: bool = False) -> Type:
-        """Type check a node in the given type context in a new frame of inferred types."""
-        while True:
-            self.binder.push_frame()
-            answer = self.accept(node, type_context)
-            changed, _ = self.binder.pop_frame(True, True)
-            self.breaking_out = False
-            if not repeat_till_fixed or not changed:
-                return answer
+    def accept_loop(self, body: Node, else_body: Node = None) -> Type:
+        """Repeatedly type check a loop body until the frame doesn't change.
+
+        Then check the else_body.
+        """
+        # The outer frame accumulates the results of all iterations
+        with self.binder.frame_context(1) as outer_frame:
+            self.binder.push_loop_frame()
+            while True:
+                outer_frame.options_on_return.append(outer_frame)  # We may skip each iteration
+                with self.binder.frame_context(1) as inner_frame:
+                    self.accept(body)
+                if not inner_frame.changed:
+                    break
+            self.binder.pop_loop_frame()
```

Comment:
Now I see that the only purpose of `Frame.changed` is to get the `changed` result from `update_from_options` out of the context manager. It doesn't even represent whether `inner_frame` changed, but rather whether `outer_frame` did, right?

Annoying that there isn't a more direct way to get this value. I think maybe it would even be better to just temporarily stick the value in a field of the binder itself rather than in the frame, or, like, pass in a dict or custom object to `frame_context` as a mutable reference that it can set return values on.